File: dotnet/System/workspace/csharp/allors.workspace/domain/runtime/IStrategy.cs
// <copyright file="Object.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace
{
    using System.Collections.Generic;
    using Meta;

    public interface IStrategy
    {
        IObject Object { get; }

        IClass Class { get; }

        long Id { get; }

        long Version { get; }

        ISession Session { get; }

        bool CanRead(IRoleType roleType);

        bool CanWrite(IRoleType roleType);

        bool CanExecute(IMethodType methodType);

        bool ExistRole(IRoleType roleType);

        object GetRole(IRoleType roleType);

        void SetRole(IRoleType roleType, object value);

        void RemoveRole(IRoleType roleType);

        object GetUnitRole(IRoleType roleType);

        void SetUnitRole(IRoleType roleType, object value);

        T GetCompositeRole<T>(IRoleType roleType) where T : IObject;

        void SetCompositeRole<T>(IRoleType roleType, T value) where T : IObject;

        IEnumerable<T> GetCompositesRole<T>(IRoleType roleType) where T : IObject;

        void AddCompositesRole<T>(IRoleType roleType, T value) where T : IObject;

        void RemoveCompositesRole<T>(IRoleType roleType, T value) where T : IObject;

        void SetCompositesRole<T>(IRoleType roleType, in IEnumerable<T> role) where T : IObject;

        T GetCompositeAssociation<T>(IAssociationType associationType) where T : IObject;

        IEnumerable<T> GetCompositesAssociation<T>(IAssociationType associationType) where T : IObject;
    }
}

File: typescript/libs/workspace/domain/system/src/lib/runtime/IStrategy.ts
import { AssociationType, Class, MethodType, RoleType } from '@allors/workspace/meta/system';
import { IObject } from './IObject';
import { ISession } from './ISession';
import { Method } from './Method';
import { UnitType } from './Types';

export interface IStrategy {
  object: IObject;

  cls: Class;

  id: number;

  session: ISession;

  canRead(roleType: RoleType): boolean;

  canWrite(roleType: RoleType): boolean;

  canExecute(methodType: MethodType): boolean;

  exist(roleType: RoleType): boolean;

  get(roleType: RoleType): unknown;

  getUnit(roleType: RoleType): UnitType;

  getComposite<T extends IObject>(roleType: RoleType): T;

  getComposites<T extends IObject>(roleType: RoleType): ReadonlyArray<T>;

  set(roleType: RoleType, value: unknown): void;

  setUnit(roleType: RoleType, value: UnitType): void;

  setComposite<T extends IObject>(roleType: RoleType, value: T): void;

  setComposites<T extends IObject>(roleType: RoleType, value: ReadonlyArray<T>): void;

  add<T extends IObject>(roleType: RoleType, value: T): void;

  remove<T extends IObject>(roleType: RoleType, value: T): void;

  removeAll(roleType: RoleType): void;

  getCompositeAssociation<T extends IObject>(associationType: AssociationType): T;

  getCompositesAssociation<T extends IObject>(associationType: AssociationType): ReadonlyArray<T>;

  method(methodType: MethodType): Method;
}
